File: android/app/src/main/java/com/travelexpense/data/model/Expense.kt
package com.travelexpense.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * 支出モデル
 */
@Serializable
data class Expense(
    val id: String,
    @SerialName("trip_id")
    val tripId: String,
    @SerialName("category_id")
    val categoryId: String? = null,
    val amount: Double,
    val currency: String,
    val description: String? = null,
    val notes: String? = null,
    @SerialName("expense_date")
    val expenseDate: String,
    val location: String? = null,
    @SerialName("payment_method")
    val paymentMethod: String? = null,
    val metadata: ExpenseMetadata? = null,
    @SerialName("created_at")
    val createdAt: String
)

/**
 * 支出メタデータ（ICカード取引情報など）
 */
@Serializable
data class ExpenseMetadata(
    @SerialName("ic_card_transaction")
    val icCardTransaction: ICCardTransaction? = null
)

/**
 * ICカード取引情報
 */
@Serializable
data class ICCardTransaction(
    @SerialName("entry_code")
    val entryCode: Int,
    @SerialName("exit_code")
    val exitCode: Int,
    val balance: Int,
    @SerialName("card_type")
    val cardType: String,
    @SerialName("raw_data")
    val rawData: String? = null
)

/**
 * 支出作成用リクエスト
 */
@Serializable
data class CreateExpenseRequest(
    @SerialName("trip_id")
    val tripId: String,
    @SerialName("category_id")
    val categoryId: String? = null,
    val amount: Double,
    val currency: String,
    val description: String? = null,
    val notes: String? = null,
    @SerialName("expense_date")
    val expenseDate: String,
    val location: String? = null,
    @SerialName("payment_method")
    val paymentMethod: String? = null,
    val metadata: ExpenseMetadata? = null
)

File: ios/TravelExpense/Models/Expense.swift
import Foundation

/// 支出モデル
struct Expense: Codable, Identifiable {
    let id: String
    let tripId: String
    let categoryId: String?
    let amount: Double
    let currency: String
    let description: String?
    let notes: String?
    let expenseDate: Date
    let location: String?
    let paymentMethod: String?
    let metadata: ExpenseMetadata?
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case tripId = "trip_id"
        case categoryId = "category_id"
        case amount
        case currency
        case description
        case notes
        case expenseDate = "expense_date"
        case location
        case paymentMethod = "payment_method"
        case metadata
        case createdAt = "created_at"
    }
}

/// 支出メタデータ（ICカード取引情報など）
struct ExpenseMetadata: Codable {
    let icCardTransaction: ICCardTransaction?

    enum CodingKeys: String, CodingKey {
        case icCardTransaction = "ic_card_transaction"
    }
}

/// ICカード取引情報
struct ICCardTransaction: Codable {
    let entryCode: Int
    let exitCode: Int
    let balance: Int
    let cardType: String
    let rawData: String?

    enum CodingKeys: String, CodingKey {
        case entryCode = "entry_code"
        case exitCode = "exit_code"
        case balance
        case cardType = "card_type"
        case rawData = "raw_data"
    }
}

/// 支出作成用リクエスト
struct CreateExpenseRequest: Codable {
    let tripId: String
    let categoryId: String?
    let amount: Double
    let currency: String
    let description: String?
    let notes: String?
    let expenseDate: Date
    let location: String?
    let paymentMethod: String?
    let metadata: ExpenseMetadata?

    enum CodingKeys: String, CodingKey {
        case tripId = "trip_id"
        case categoryId = "category_id"
        case amount
        case currency
        case description
        case notes
        case expenseDate = "expense_date"
        case location
        case paymentMethod = "payment_method"
        case metadata
    }
}

/// 支出更新用リクエスト
struct UpdateExpenseRequest: Codable {
    let categoryId: String?
    let amount: Double?
    let currency: String?
    let description: String?
    let notes: String?
    let expenseDate: Date?
    let location: String?
    let paymentMethod: String?

    enum CodingKeys: String, CodingKey {
        case categoryId = "category_id"
        case amount
        case currency
        case description
        case notes
        case expenseDate = "expense_date"
        case location
        case paymentMethod = "payment_method"
    }
}
